File: rust/src/parse/parsers/csv.rs
use crate::parse::*;

pub fn csv_parser<'a>() -> Parser<'a, Vec<Vec<&'a str>>> {
    let double_quotes = || string("\"");
    let single_quotes = || string("'");

    let token = regex("[^\"]+").wrap(double_quotes(), double_quotes())
        | regex("[^']+").wrap(single_quotes(), single_quotes())
        | regex(r"[^,\r\n]+")
        | string("").look_ahead(string(","));

    let delim = string(",");

    let line = token.sep_by(delim, ..).skip(regex(r"\s")).trim_whitespace();
    let csv = line.many(1..);

    csv.trim_whitespace()
}

File: typescript/src/parse/parsers/csv.ts
/**
 * Combinator CSV parser — built from parse-that primitives.
 * Handles double-quoted, single-quoted, and unquoted fields.
 */
import { regex, any, string, Parser } from "../index.js";

const delim = string(",").trim();
const doubleQuotes = string('"');
const singleQuotes = string("'");

const token = any(
    regex(/[^"]+/).wrap(doubleQuotes, doubleQuotes),
    regex(/[^']+/).wrap(singleQuotes, singleQuotes),
    regex(/[^,]+/)
);

const line = token.sepBy(delim).trim();

/** Combinator CSV parser — returns array of rows, each an array of string fields. */
export const csvParser = line.many();
